File: 34.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.empty())
            return vector<int>{-1,-1};
        int l, r;
        int start, end;
        start = 0;
        end = nums.size();
        while(start < end){
            int mid = (start + end)/2;
            if(nums[mid]>=target){
                end = mid;
            }else{
                start = mid+1;
            }
        }
        l = start;
        end = nums.size();
        while(start < end){
            int mid = (start+end)/2;
            if(nums[mid]<=target){
                start = mid+1;
            }else{
                end = mid;
            }
        }
        r = start;
        return l==r? vector<int>{-1,-1}:vector<int>{l,r-1};
    }
};

File: golang/34/34.go
package main

func searchRange(nums []int, target int) []int {
	left := left_bound(nums, target)
	right := right_bound(nums, target)
	return []int{left, right}
}

func left_bound(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] == target {
			right = mid - 1
		} else if nums[mid] < target {
			left = mid + 1
		} else if nums[mid] > target {
			right = mid - 1
		}
	}

	if left >= len(nums) || nums[left] != target {
		return -1
	}

	return left
}

func right_bound(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] == target {
			left = mid + 1
		} else if nums[mid] > target {
			right = mid - 1
		} else if nums[mid] < target {
			left = mid + 1
		}
	}

	if right < 0 || nums[right] != target {
		return -1
	}
	return right
}
